[MRG] Interactive annotations.

jaeilepp opened the pull request:
Closes https://github.com/mne-tools/mne-python/issues/3467

I thought that the right way of doing this is to have 'modes' for different behaviors, so you don't have to plot everything again to select bad channels. Now the annotation mode is toggled by pressing 'a'.

Still to do:
- [x] modification,
- [x] deletion,
- [x] (maybe) text edit for description (mpl does not have textboxes before next release),
- [x] ~~nicer visuals,~~
- [x] docs/tests.

![screenshot_2016-12-07_13-09-02](https://cloud.githubusercontent.com/assets/3456414/20967589/8adb37d8-bc80-11e6-94e0-f980788ef1eb.png)

Wdyt?

mmagnuski commented:
I was waiting for this to happen - happy it's happening now :)
I will tests it today, but I like it being an annotation mode.

BTW - sometimes (touchpad or when annotation will need to span more than one 'screen') instead of dragging it is more convenient and precise to click for start and then shift-click for stop.

agramfort commented:
very excited about this feature !

jaeilepp commented:
> BTW - sometimes (touchpad or when annotation will need to span more than one 'screen') instead of dragging it is more convenient and precise to click for start and then shift-click for stop.

I believe the control keys still don't work on MAC. It is always possible to extend the view though. Perhaps the modifying feature will eventually overcome the problem with precision.

jaeilepp commented:
Now the addition and removal work (surprisingly smoothly IMO) and maybe now would be a good moment to get some suggestions on the visuals and the usability. The annotation mode is toggled by pressing 'a'. Annotations are made by click/dragging with left mouse button, modified by dragging the edge of existing annotation and deleted with a right click. These instructions could be added to the annotation dialog, if we eventually want it to stay. The mode is exited by closing the annotation dialog or pressing 'a'.

Eric89GXL commented:
@mmagnuski do you have time to test it out?

@hoechenberger that heart emoji makes me think you might be interested in trying it, too :)

mmagnuski commented:
Yes, I played with it and it works well, I like it :) Just two things:
* when I am still in annotations mode and close the whole figure the annotation figure should be closed automatically too
* when I drag a limit of one of the annotations so that it overlaps with another annotation of the same type - they should be merged (not necessarily in this PR).

Another, loosely related question: is the sluggish behavior of plotting (not only the annotations) a general matplotlib limitation or in some part limitation of the current implementation (partial functions passing around dict etc.)?

hoechenberger commented:
@Eric89GXL, I'm currently mostly using MNE only on a headless server (with lots of memory and CPU power) via the Jupyter notebook. Many interactive features work if the IPython magic `%matplotlib notebook` is executed. I would assume that the approach with opening a second window etc. wouldn't work in this environment -- I cannot test until monday, though!

agramfort commented:
would it be possible to annotate at the channel level?

jaeilepp commented:
> would it be possible to annotate at the channel level?

Perhaps. That would mean changes to the Annotations class. Maybe another PR.

mmagnuski commented:
@agramfort @jaeilepp Would be nice (some time later) - and along with annotations for epochs it could play nicely with autoreject.

mmagnuski commented:
BTW - text description (and maybe - choosing color :) ) would also be good for another PR from my point of view.

jaeilepp commented:
> Another, loosely related question: is the sluggish behavior of plotting (not only the annotations) a general matplotlib limitation or in some part limitation of the current implementation (partial functions passing around dict etc.)?

I believe the raw plotter is quite lean in the sense that it does not do a lot of runtime processing. With the epochs plotter there are some extra checks made, and it's using line collections, so there is a small difference in speed. Mostly the speed is proportional to the amount of data plotted.

mmagnuski commented:
@jaeilepp ok, thanks! I was just comparing the speed to matlab (but not the eeglab plotter - which at least some time before was very inefficient) and was wondering whether it is matplotlib that slow or maybe something can be improved (anyone checked the gr backend btw or gr itself?).

Eric89GXL commented:
The style Travis build was still unhappy so I killed all but the 2.7-full one

jaeilepp commented:
Ok. This is ready for review. The modifying is a bit laggy because it uses the hover listener, but it's still usable on my computer and it only affects the annotation mode. If it's too slow for someone, maybe we can switch it off and do it some other way...

mmagnuski commented:
No, I was just curious about the speed limitations of matplotlib. Modification has noticable lag, but its rather esthetical thing - it is still convenient.

Eric89GXL commented on tutorials/plot_artifacts_correction_rejection.py:
```diff
@@ -133,8 +133,8 @@
-# It is also possible to draw bad segments interactively using ``raw.plot``
-# (see :ref:`tut_viz_raw`).
+# It is also possible to draw bad segments interactively using
+# :func:`raw.plot <mne.viz.plot_raw>` (see :ref:`tut_viz_raw`).
```
Or
```
:meth:`raw.plot <mne.io.Raw.plot>`
```
?

Eric89GXL commented on mne/cov.py:
```diff
@@ -1197,7 +1197,7 @@ def _unpack_epochs(epochs):
 
 def _get_ch_whitener(A, pca, ch_type, rank):
-    """"Get whitener params for a set of channels."""
+    """Get whitener params for a set of channels."""
```
FYI these changes are also in #3831

jaeilepp commented:
The hover event seems to be very buggy on old mpl. On 1.3 it messes up everything and while on 1.4 it works, it occasionally throws an error, so I disabled the modifying from mpl < 1.5.

jaeilepp commented:
This is ready for review. The text box would be nice, but I have no clear idea how it could be done reasonably with matplotlib.

agramfort commented:
ok interactivity works like a charm for me.

in terms of usability current code only allows me to mark bad segments.

I may want to annotate some signal features such as alpha burst, sleep
spindles, k-complex but also distinguish between the different type of
artifacts.

do you see a simple way to do this?

jaeilepp commented:
I believe the CIs will be happy about it now. Now the descriptions are also modifiable. Text boxes are not yet available in mpl, so the dialog now simply listens to key presses and updates the description accordingly. Suggestions/comments about the usability and visuals are welcome.

jaeilepp commented:
![screenshot_2017-01-12_15-26-23](https://cloud.githubusercontent.com/assets/3456414/21893499/8bb2f26e-d8db-11e6-97ce-28901f2e4cfa.png)

Eric89GXL commented:
That looks cool. Just from looking, one suggestion: color the text and circle-outlines based on the color that is used in the annotation-plotting window.

Maybe @jasmainak you could looks since you've been doing some annotation-related things lately?

jasmainak commented:
looks cool ... I'll take a look tomorrow morning

jasmainak commented:
A couple of usability feedbacks:

1. Why is some stuff in green and others in red? It's not obvious to me. I think it should all be the same color or all should be different colors.
2. If I write the name of the annotation, then make a segment, then I can go back and edit the text again. And now if I make a segment, it's marked with the new text but I see only one item in the pop up menu. This is confusing
3. If I make an annotation and close the raw.plot(). Now I fire up raw.plot() again and use 'a' to start annotation mode. Now, I can't delete the previous annotations. It gets confusing which ones I can delete and which ones I can't ...

Visual feedback:

1. I think "Annotate bad data by dragging on top ... window" would look better if it's not in bold.

Feature request:

1. Can I have the time stamp of the vertical line highlighted? It can be above the "Help" button. This will allow users to make accurate annotations ...

Hope this much feedback is enough for now ;)

jaeilepp commented:
I think I've mostly addressed @jasmainak comments.

jasmainak commented:
@jaeilepp more comments:

1. I cannot make segments if the text in the annotations is empty
2. It's a bit unintuitive that the new labels are getting randomly at any position rather than at the top rather than the bottom. See this for example
![sdfdfd](https://cloud.githubusercontent.com/assets/15852194/22035309/bf2e7b96-dcef-11e6-9dca-78a1ae751928.png)
3. I think it's not super clear at the moment how to add a new label. You have to check the label and then make a segment to add a new empty label. Maybe just pressing enter will add a new label? And maybe also put it in the help text how to do this ...
4. The first label starts with "BAD_" but the remaining labels start with just "\_". Perhaps make it consistent to be "BAD_"?

jaeilepp commented:
@jasmainak I changed the dialog a bit. See how it is now.

![screenshot_2017-01-18_14-44-35](https://cloud.githubusercontent.com/assets/3456414/22066308/aebf3cd0-dd8c-11e6-8e0a-29a4803f1bef.png)

jasmainak commented:
@jaeilepp much better!

There is still the problem of the empty label though. You cannot draw an annotation when the label is empty. Perhaps don't allow the user to "add" if you cannot draw an annotation with empty label?

Also, I noticed that you allow overlapping annotations. Is this on purpose? But when they are overlapping, I cannot select the edge of the segment which is below ...

jasmainak commented:
Pretty neat that if I drag the edges of one annotation on to another annotation of the same type, they merge together ... but again not sure if it's a "feature" or a "bug"

jasmainak commented:
I had a chat with @jaeilepp IRL and we agreed that handling overlapping annotations was not worth the added code complexity. Good from my end.

Anyone else wants to look? Otherwise it has my +1

jasmainak commented:
@jaeilepp don't forget to update `whats_new`!

Eric89GXL commented on mne/annotations.py:
```diff
@@ -84,6 +84,36 @@ def __init__(self, onset, duration, description,
+        """
+        self.onset = np.delete(self.onset, idx)
+        self.duration = np.delete(self.duration, idx)
+        self.description = np.delete(self.description, idx)
```
need tests for these?

jaeilepp replied:
Added to ``test_annotations.py``.

Eric89GXL commented on mne/viz/tests/test_raw.py:
```diff
@@ -75,22 +76,10 @@ def test_plot_raw():
-        fig.canvas.key_press_event('escape')
+        for key in ['escape', 'down', 'up', 'right', 'left', 'o', '-', '+',
+                    '=', 'pageup', 'pagedown', 'home', 'end', '?', 'f11',
+                    'escape']:
```
escape is in there twice, is it intentional?

Eric89GXL commented on mne/viz/tests/test_raw.py:
```diff
@@ -134,6 +123,57 @@ def test_plot_raw():
         plt.close('all')
 
 
+@requires_version('matplotlib', '1.2')
```
our minimum requirement is 1.3 already so you don't need this

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -261,10 +263,10 @@ def _get_help_text(params):
+    text.append(u'\u2190 : \n')  # left arrow
+    text.append(u'\u2192 : \n')  # right arrow
+    text.append(u'\u2193 : \n')  # down arrow
+    text.append(u'\u2191 : \n')  # up arrow
```
why not actually put the unicode characters in the file?

https://unicode-table.com/en/sets/arrows-symbols/

jaeilepp replied:
Does not work for me.

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -612,6 +616,13 @@ def _radio_clicked(label, params):
 
+def _get_active_radiobutton(radio):
+    """Helper to find out active radio button."""
+    # XXX: In mpl 1.5 you can do: fig.radio.value_selected
```
you can do a `try/except` then here, no?

jaeilepp replied:
This returns an index and ``value_selected`` returns the text, so I would have to map it to the index. I think it's cleaner like this.

Eric89GXL commented on tutorials/plot_visualize_raw.py:
```diff
@@ -38,8 +38,18 @@
+# viewed at once.
+#
+# You can enter annotation mode by pressing ``a`` key. In annotation mode you
+# can mark segments of data (and modify existing annotations) with left mouse
```
the left mouse

Eric89GXL commented on tutorials/plot_visualize_raw.py:
```diff
@@ -38,8 +38,18 @@
+# You can enter annotation mode by pressing ``a`` key. In annotation mode you
+# can mark segments of data (and modify existing annotations) with left mouse
+# button. You can use the description of any existing annotation or create a
+# new description by typing when annotation dialog is active. Notice that
```
the annotation dialog

Eric89GXL commented on tutorials/plot_visualize_raw.py:
```diff
@@ -38,8 +38,18 @@
+# can mark segments of data (and modify existing annotations) with left mouse
+# button. You can use the description of any existing annotation or create a
+# new description by typing when annotation dialog is active. Notice that
+# description starting with keyword ``'bad'`` means that the segment will be
```
the description

Eric89GXL replied:
the keyword

Eric89GXL commented on tutorials/plot_visualize_raw.py:
```diff
@@ -38,8 +38,18 @@
+# button. You can use the description of any existing annotation or create a
+# new description by typing when annotation dialog is active. Notice that
+# description starting with keyword ``'bad'`` means that the segment will be
+# discarded when epoching the data. Existing annotation can be deleted with
```
Existing annotations

Eric89GXL commented on tutorials/plot_visualize_raw.py:
```diff
@@ -38,8 +38,18 @@
+# new description by typing when annotation dialog is active. Notice that
+# description starting with keyword ``'bad'`` means that the segment will be
+# discarded when epoching the data. Existing annotation can be deleted with
+# right mouse button.  Annotation mode is exited by pressing ``a`` again or
```
the right mouse button

jaeilepp commented:
After discussion IRL with @agramfort I changed the colors to of the circle to gray and the edgecolor indicates the color of the annotation. Also, now the latest addition to the label list is automatically activated.
![ann_dialog](https://cloud.githubusercontent.com/assets/3456414/22371113/6b0247fc-e495-11e6-9667-f6ee4f94447d.png)

jasmainak commented:
oh yes! Much much nicer!

agramfort commented:
try with this: https://github.com/agramfort/mne-python/commit/d73e3e0263f21c579c756177bf2f404ec041f6c2

but for some reason now text of label overlaps with radio button

do you understand why?

jaeilepp commented:
This is how it looks now
![buttons](https://cloud.githubusercontent.com/assets/3456414/22375112/18f81072-e4a9-11e6-9354-01fe0763f8e1.png)

agramfort commented:
is it possible to add some space between button and the text?

also can you do:

```
fig = figure_nobar(figsize=(4.5, 2.75 + len(labels) * 0.75))
fig.patch.set_facecolor('white')
```

so figure is less wide and the background is white.

thx @jaeilepp

agramfort commented:
@jaeilepp I pushed a cosmit directly to your branch

jaeilepp commented:
![radio](https://cloud.githubusercontent.com/assets/3456414/22419098/2913c566-e6dc-11e6-9a73-47134b00a2de.png)
LGTM

agramfort commented:
thx @jaeilepp !
